### ISSUE TITLE
Stutter when resizing very large typography

### ISSUE THREAD

**Issue by Arecsu** [photopea/photopea#4889]
Hi! Thank you so much for your work on Photopea. It's truly an amazing project :)

I've been doing some tests on it, and discovered this issue:

https://user-images.githubusercontent.com/12679098/183078618-48dbceb4-ba9a-43e9-a595-eb3b81e3d688.mp4

- Stutter when resizing very large typography with transform
- Changing the scale algorithm while transforming it, although I'm sure if it really does something give that it's a vector graphic, it blacks out the entire layer

I suspect that it tries to render the vector graphic at 100% the quality, which of course is desirable. However, when transforming, I think the priority should be speed. Maybe rendering at lower quality, blurry or pixelated as the typography gets bigger based on document resolution+layer resolution/size? And at the moment the user finished moving the mouse (transforming it), it could start processing the preview of the vector at full quality.

I really think sacrificing some quality WHILE transforming something is a very good UX approach. If I recall correctly, that's what Affinity Photo does. When I use Photoshop, I always find the stutter and low FPS of the transform function (for both vector and raster graphics) to be very annoying when doing design work. It's a very demanding process, but maybe with the suggestions of rendering at lower quality while transforming takes place and 100% quality when user stopped moving, rotating or resizing the object should result in a much better UX.

Image specs:
- 6000x6000px, 8-bit

My specs:
- Brave browser, latest version. Haven't tested on other browsers.
- nvidia 2070 Super, amd 3700x, Windows 11
- 3840x2160px screen size, scaling 150%

**Comment by photopea** [comment]
You are telling Photopea to redraw 36 million pixels several times per second.

That is the best performance you can get out of your computer. If we rendered it at a lower resolution, many people would be confused, as they dont expect the quality to drop while using the Free Transform.

You can get a better performance by getting a better CPU. We are planning to make a vector editor, which does not need to rasterize every pixel.

**Comment by Arecsu** [comment]
Sorry, I think there's a misunderstanding.

I wasn't suggest to drop the final quality of the Free Transform Tool. I was suggesting to only drop it *while* the user is performing the action. Once the user stopped moving, rotating or scaling the layer, then the process of rendering it at 100% quality should begin.

This is Affinity Photo, for reference. They are doing that technique I'm suggesting it so the usage of the Transform tool feels smooth. They're using it even for the canvas Zoom.

https://user-images.githubusercontent.com/12679098/183112820-6ece9cbc-3246-49fa-a2b3-d2e7aa62a9ba.mp4

The drop in quality is very low though here, barely noticeable in the video. But I can totally see it in my computer. In certain effects it drops even lower to maintain good fps. Quality, once the user stopped doing the transformation, will always be rendered at 100%.

**Comment by photopea** [comment]
I can not see any drop in quality in your video.

Are you sure they are doing raster graphics and not vector graphics? Could you add a drop shadow behind your text?

**Comment by Arecsu** [comment]
Yes. The text tool is a vector graphic. 

As you suggested, adding more effects to the text forced Affinity to show more clearly what I've been suggesting:

https://user-images.githubusercontent.com/12679098/183144028-18d07864-ed1c-4413-96f7-6caca16e8702.mp4

I've also set the canvas to a really high resolution (80.000 x 40.000 px) and noticed another thing: the performance is exactly the same in a much lower resolution, like 4000 x 2000px. 

I think what Affinity does is render the current view only at the exact amount of pixels being displayed on screen. It wouldn't make sense to render 80.000px when there's just 2000px in the screen. It's just way more optimal. When zoom-in, it renders the corresponding part of the image being displayed.

That way, combined with the transform "optimizations" I've showed in the video, would translate in huge perceived performance gains and better user experience. I put quotes in optimizations because it really isn't.
